clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java

```diff
@@ -634,12 +634,12 @@ public boolean matches(ClientRequest request) {
 
 
     @Test
-    public void testExcludeInternalTopicsConfigOption() { 
+    public void testExcludeInternalTopicsConfigOption() {
         subscriptions.subscribe(Pattern.compile(".*"), rebalanceListener);
 
         metadata.update(TestUtils.singletonCluster(TopicConstants.GROUP_METADATA_TOPIC_NAME, 2), time.milliseconds());
 
-        assertFalse(subscriptions.partitionAssignmentNeeded());
+        assertTrue(subscriptions.partitionAssignmentNeeded());
```

Comment:
Here's the code that should get triggered when the metadata is refreshed: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L161. One definite problem is that we didn't request a metadata update when the pattern changed. That would have meant that we wouldn't rebalance until the next periodic metadata refresh (which occurs every 5 minutes by default). I think the question is whether that logic works correctly or if there is another problem.